### PR TITLE
Update IncrementingPollingCounter only on Timer thread

### PR DESCRIPTION
**Pull request by eterekhin**
My idea is to call `_totalValueProvider()` only on Timer thread . It should fix deadlocks like https://github.com/dotnet/runtime/issues/93175  since the Timer thread doesn't hold ```EventListener.EventListenersLock```

### CODE REVIEW

**Comment by eterekhin** [comment]
@dotnet-policy-service agree

**Comment by eterekhin** [comment]
@davmason, Hey! May I ask you too look at this?

**Comment by davmason** [comment]
Added other people from dotnet-diag for review

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs
Comment (davmason) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs:49]: The name `ResetStatistics` doesn't make sense to me here, it's not actually resetting any statistics. It just causes an extra update on the next timer poll.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs
Comment (davmason) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs:64]: @noahfalk just calling Update twice in a row here doesn't feel right

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs
Comment (davmason) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs:206]: We also have the same issue for IncrementingEventCounter on line 202, right?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs
Comment (eterekhin) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs:64]: Thank you for review!
While working on this issue I learned that we do reset to fix counter values on first reporting and it seemed okay to me to do 2 calls in a row in this case (https://github.com/dotnet/coreclr/pull/25739).

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs
Comment (eterekhin) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs:206]: ```IncrementingEventCounter.UpdateMetric``` call looks safe for calling under ```EventListener.EventListenersLock``` since it just sets ```_prevIncrement``` value
But ```IncrementingPollingCounter``` call ```_totalValueProvider``` action (an arbitrary action passed via ```IncrementingPollingCounter``` ctor) which can be dangerous for us

We can unify the reset logic for all ```IncrementingEventCounter```, ```IncrementingPollingCounter```, ```EventCounter``` to not update metrics under ```EventListener.EventListenersLock```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs
Comment (eterekhin) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs:49]: If we decide to go with this approach I would be happy to find a different name and unify this reset logic for ```IncrementingEventCounter```, ```IncrementingPollingCounter``` and  ```EventCounter```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs
Comment (noahfalk) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs:64]: Two calls in a row likely means that all IncrementingPollingCounters would return 0 (or near 0) on the first report because very little time has passed for the counter to be incremented. dotnet/coreclr#25739 was attempting to ensure the first value reported for counters was meaningful. In the case of IncrementingPollingCounter that meant querying the value once, then waiting the duration of one sampling interval and polling again. The reported difference should represent the change in value during one whole sampling time interval.

I think the idea to move the query to the timer thread is the right direction, but we still want it to happen as soon as possible rather than delaying it until the end of the next sampling interval. I think you could accomplish that doing something like this...

Add a new static list:
```C#
class CounterGroup
{
    ...
    private static List<CounterGroup>? s_needsResetCounterGroup;
```

In EnableTimer() add counter groups to that list and don't call ResetStatistics():
```C#
if(!s_needsResetCounterGroup.Contains(this))
{
    s_needsResetCounterGroup.Add(this);
}
// notify the polling thread that the polling interval may have changed and the sleep should
// be recomputed
s_pollingThreadSleepEvent!.Set();
```

In PollForValues() snapshot the list and invoke ResetCounters() on each item in it:
```C#
List<CounterGroup> groupsToReset = new List<CounterGroup>();
while(true)
{
    ...
    lock(s_counterGroupLock)
    {
        if(s_needsResetCounterGroup.Count > 0)
        {
            groupsToReset = s_needsResetCounterGroup;
            s_needsResetCounterGroup = new List<CounterGroup>();
        }
        ...
    }

    foreach(var group in groupsToReset)
    {
        group.ResetStatistics();
    }
    groupsToReset.Clear();
    ...
}
```

And in DisableTimer() remove the counter group if it happens to still be there. This should be a rare case where the timer is enabled and then disabled extremely quickly.
```C#
DisableTimer()
{
    ...
    s_needsResetCounterGroup?.Remove(this);
}
```

I expect in the typical case the polling thread will wake up once because s_pollingThreadSleepEvent was signaled and it will run all the calls to ResetStatistics() at that point. Then at the end of polling interval it will wake up again, observe that polling time has elapsed and start invoking all the EventCounter OnTimer callbacks. Its not perfect as we aren't guaranteed the PollingThread will wake up quickly after the signal, but if the app is so CPU constrained it can't reliably wake up the polling thread on time this won't be the only measurement that gets skewed.

Comment (eterekhin) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs]: Hey, @noahfalk! Thanks a lot for review! Please take a look at my changes

In EnableTimer method I had to add waiting for reset to finish. Without this waiting, tracing tests failed https://dev.azure.com/dnceng-public/public/_build/results?buildId=762610&view=results11. 
I looked at TestIncrementingEventCounter and as far as I understand it failed because after creating SimpleEventSource we immediately increment the counter from the test. At the same time our reset is scheduled to run on Timer thread, its execution will eventually overwrite the values written from the test

I would also add exception handling in PollForValues, to unblock a thread waiting in EnableTimer if there's an exception in Timer thread.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs
Comment (noahfalk) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs:186]: I believe blocking here would recreate the deadlock conditions we are trying to avoid. When running the example described in https://github.com/dotnet/runtime/issues/93175, I would expect:

Thread A - Main thread calling Threadpool.GetMaxThreads() - Acquires static constructor lock, waits on EventListenerLock
Thread B - ETW callback thread calling EnableTimer - Acquires EventListenerLock, waits on this new event
Thread C - Timer thread calling Threadpool.get_CompletedWorkItemCount() - Waits on static constructor lock, never signals the new event

In total A waits on B waits on C which waits on A again.


I see what you mean about the tests calling WriteMetric() immediately after creating the EventSource and assuming that the data would be captured. https://github.com/dotnet/runtime/blob/main/src/tests/tracing/eventcounter/eventcounter.cs#L121

While this behavior probably isn't a promise we ever intentionally wanted to make its possible other code has taken a dependency on it too and its probably not too hard to preserve it. I'm thinking now we could reset all counter types except IncrementingPollingCounter synchronously in EnableTimer() and only do the callback that initializes IncrementingPollingCounter asynchronously on the Timer thread. That looks like it would preserve as much of the existing synchronous reset behavior as we can while still addressing the deadlock. You could resume using ResetCounters() for synchronous reset behavior and then ResetCounterGroupIfNeeded() could be repurposed into something like InitIncrementingPollingCounters().

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs
Comment (noahfalk) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs]: Happy to :) I think this current version of the code still has a modified form of the original deadlock and you will need to get rid of blocking on that new event you added to resolve it. I commented a little more detailed inline.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs
Comment (eterekhin) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs]: @noahfalk, you're right, I missed that case, sorry :( 
Thank you for pointing it out!

I pushed a version with resetting only IncrementingPollingCounters on the Timer thread, I decided to put IncrementPollingCounters in a static list. I considered static CounterGroup field as alternative, but, in this case  CounterGroup._counters list can be changed within releasing s_counterGroupLock in EnableTimer method and taking it in the Timer thread.
Could you please take a look?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs
Comment (noahfalk) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs:307]: Nit: The code as-is looks like it would work, but I think a swap would simplify it a bit:
```suggestion
                    countersToReset = s_needsResetIncrementingPollingCounters;
                    s_needsResetIncrementingPollingCounters = [];
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs
Comment (noahfalk) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs:158]:
```suggestion
                // Schedule IncrementingPollingCounter reset and synchronously reset other counters
                HandleCountersReset(); 
```
Nit: The thread may already be running

**Review by noahfalk — COMMENTED** [review]
This looks good to me - I suggested just a few minor tweaks.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs
Comment (eterekhin) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs:308]: Thank you! 

I addressed it, just decided to leave ```if (s_needsResetIncrementingPollingCounters.Count > 0)``` there to not create a new list on each iteration
Take a look please ;)

**Review by noahfalk — APPROVED** [review]
LGTM, thanks for your work on this @eterekhin!

**Comment by noahfalk** [comment]
@davmason - did you want to take another look?
@eterekhin - I assume you are ready for this to merge if @davmason is also happy with it?

**Review by davmason — APPROVED** [review]
LGTM

**Comment by eterekhin** [comment]
Thanks a lot for helping with this fix, folks!
@davmason, @noahfalk Could you please merge it? I guess the "Merge" button is hidden for me
image

**Comment by tarekgh** [comment]
@noahfalk do we need a breaking change doc for this? if so, please file it through https://github.com/dotnet/docs/issues/new?assignees=gewarren&labels=breaking-change%2CPri1%2Cdoc-idea&projects=&template=02-breaking-change.yml&title=%5BBreaking+change%5D%3A+ and mark this PR with breaking change label?

**Comment by noahfalk** [comment]
Would we typically write breaking change docs when we change an implementation choice that wasn't part of the original contract? I don't think of this as a breaking change in the formal sense but its true that some people may have taken dependencies on the threading behavior anyways.

**Comment by tarekgh** [comment]
We usually document any change that can cause breaking to the user. If this change is possible to break any behavior, I suggest documenting it, but it is your call if you think otherwise.

**Comment by noahfalk** [comment]
Cool, I'll file it.

**Pull request merged**
